### DISCUSSION TITLE
How to control DPI for `realistic_solder_mask: true` in `pcb_print` output?

### DISCUSSION THREAD

**Discussion by cdwilson** [INTI-CMNB/KiBot#259] [Q&A]
I'm wondering if it's possible to control the DPI used for the `realistic_solder_mask: true` in `pcb_print` output.  When I generate a PDF with this option enabled, the resulting PDF looks very pixelated (not sure if I'm doing something wrong):

<img width="840" alt="image" src="https://user-images.githubusercontent.com/14631/186040194-a692edca-680b-4a8e-8270-9a6213fb4753.png">

For comparison, when I generate a PNG with `pcbdraw` output, I can control the DPI so that the image looks sharp:

![image](https://user-images.githubusercontent.com/14631/186040434-2b1675be-06d2-4864-a924-069a2f8b88fa.png)

**Comment by set-soft** [comment]
Hi @cdwilson !
Did you try other output formats?
I think this is a problem with the SVG to PDF conversion. Take a look at the edge, it isn't pixelated at all. I noticed some problems, never that pixelated.

**Reply by cdwilson** [reply]
> I think this is a problem with the SVG to PDF conversion.

I think you're right.  I don't know how this conversion works internally, but I just tried exporting SVG and it looks perfect (since it's a vector format):

<img width="627" alt="image" src="https://user-images.githubusercontent.com/14631/186194654-f04f4f0c-4360-4eb1-8811-93e0b80e0c0c.png">

Since SVG and PDF are both vector formats, it seems like it should be possible to create the PDF without rasterizing the image in the conversion.  Do you know why it's getting rasterized and if it's possible to do a lossless conversion straight from SVG to PDF?

**Reply by cdwilson** [reply]
> I noticed some problems, never that pixelated.

I think it might be because this board is very small (1x1 inch).

FYI, I also tried PNG and I see the same pixelation:

<img width="513" alt="image" src="https://user-images.githubusercontent.com/14631/186195556-9839de50-1552-426e-a88f-f0594bd8cef8.png">

**Reply by set-soft** [reply]
Hi @cdwilson !
Can you upload a similar file? You can remove most of the components and tracks. I can try a couple of things. The current main problem is that adjusting the resolution will also change the size of the PDF size.

**Reply by cdwilson** [reply]
See attached [ModST_Mouse.kicad_pcb.zip](https://github.com/INTI-CMNB/KiBot/files/9404324/ModST_Mouse.kicad_pcb.zip) (the design will eventually be published open source on GitHub, but I'm just not done with it yet...)

**Reply by cdwilson** [reply]
Hi @set-soft, just FYI, I did some additional experiments using Inkscape to export SVG soldermask layers to PNG.  I was able to get really good results for the realistic soldermask by setting the DPI in Inkscape to a really high value (like 1200).

The PNG output for `pcb_print` does not currently have a `dpi:` option, but it does have a `png_width:` option which I used as a proxy for DPI in my experiments.  For example, here's an example PNG generated by KiBot directly with `png_width: 13200`, which is equivalent to DPI=1200:
![ModST_Mouse_v1_FCD-000037_Rev_B_assembly_page_11](https://user-images.githubusercontent.com/14631/188517263-c8330d23-58d1-46a1-9a13-78b8cbf80aa7.png)

> Since SVG and PDF are both vector formats, it seems like it should be possible to create the PDF without rasterizing the image in the conversion.

If it is not possible to achieve lossless SVG-to-PDF conversion without an intermediate rasterization step, one alternative option would be to add a `dpi:` option to the `pcb_print` output type.  This would allow the user to control the tradeoff between the output resolution vs. output file size for any outputs where rasterization is happening (i.e. PDF, PNG, etc).  BTW, I think it's helpful to have both `dpi:` and `png_width:` options (mutually exclusive)--in some cases, users may want all PNGs to have the same DPI resolution even if the output images have different widths, while other users may want all PNG images to have the same width, irrespective of DPI resolution.  What do you think?

**Reply by set-soft** [reply]
The main problem messing with the DPIs is that you destroy the scales.
The code uses 72 dpi because this is what an SVG point is supposed to represent. If we change it to 300 dpi the page size goes from US Letter to 1164 x 900 mm ... not a good idea.
I'm investigating options.

**Reply by set-soft** [reply]
368488906000ecec375a3ad272519a69c8a43d98 should solve it.

Note that this problem is only relevant for the solder mask. I think RSVG doesn't know a mechanism to map the solder mask into some vectorized object in the PDF, so it generates a rasterized object. Perhaps is too complex and the tool just uses a bitmap.

I changed most of the conversion chain, now is slower, but we have control over the resolution.
SVGs are generated directly, all the other formats are first converted to PDF, then we join the PDF files using internal Python code, during this step I can adjust the page size (broken during the SVG -> PDF conversion because we changed the DPI) and finally we convert the PDF into the output format using Ghostscript (can handle PS, EPS and PNG without problems).
So now a `dpi` option controls the resolution (defaults to 360, 5 times more than the default).

In addition using `png_width: 0` the PNG width is adjusted according to the DPI value.

**Reply by cdwilson** [reply]
Hi @set-soft, thanks for looking into this.

I'm seeing some issues when trying to generate the realistic soldermask output.  For example,

`realistic_solder_mask: false`:
![plot_png_no_realistic_soldermask_assembly_page_01](https://user-images.githubusercontent.com/14631/188962237-9b96d0e4-8df7-42b0-844d-2d27f940354e.png)

`realistic_solder_mask: true`:
![plot_png_assembly_page_01](https://user-images.githubusercontent.com/14631/188962252-c7fe95b5-f913-41c1-a292-21125cd04185.png)

Can you checkout this example project?  I'm not sure if there is a bug or something I'm doing wrong in the config.

```
git clone --branch test-soldermask-changes git@github.com:cdwilson/test-kibot.git
cd test-kibot
make
# compare realistic vs. non-realistic soldermask outputs in _build/
```

**Comment by set-soft** [comment]
Hi @cdwilson !
I used your test repo and got: 
![plot_png_no_realistic_soldermask_assembly_page_01](https://user-images.githubusercontent.com/22823627/189086888-4eb09986-e927-4b6a-8b07-b8e07ea6fb7e.png)
![plot_png_assembly_page_01](https://user-images.githubusercontent.com/22823627/189086990-f7b16659-6ace-4d53-8a52-00d3fef5ebf6.png)

Here is the full output for `kibot -d results`

[result.zip](https://github.com/INTI-CMNB/KiBot/files/9525616/result.zip)

I got the same using my system and the `dev_k6` image, I also tried `make release` and got the same.

**Reply by cdwilson** [reply]
Hi @set-soft, thanks for testing this out.

I pulled the latest docker image, and now I'm getting the same results as your [result.zip](https://github.com/INTI-CMNB/KiBot/files/9525616/result.zip).  However, if you open the `plot_pdf.pdf` in that zip, you'll see that the soldermask is not being drawn correctly in the PDF.  Here's what I see when I open the PDF:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/14631/189195823-c4f7aee9-bd80-4e74-b98f-1ac5e85c2c67.png">

Also, I tried this same docker image on a different design, and I'm getting strange results where only part of the soldermask is being drawn:

PDF: [ModST_Mouse_v1_FCD-000037_Rev_B_plots.pdf](https://github.com/INTI-CMNB/KiBot/files/9529608/ModST_Mouse_v1_FCD-000037_Rev_B_plots.pdf)
PNG: [ModST_Mouse-F_Mask](https://user-images.githubusercontent.com/14631/189196220-948b6dfb-5bd4-444c-8cb1-455bafabe9b2.png)
PNG: [ModST_Mouse-B_Mask](https://user-images.githubusercontent.com/14631/189196254-ea3c76d6-b24b-4c8f-959d-524c5dd200a9.png)
SVG: [ModST_Mouse-F_Mask](https://user-images.githubusercontent.com/14631/189196680-e89e5973-e92f-4abd-957f-6124a2daf694.svg)
SVG: [ModST_Mouse-B_Mask](https://user-images.githubusercontent.com/14631/189196726-1b71bb68-84fc-4822-a73e-83dbc1c6e79d.svg)

**Reply by set-soft** [reply]
Hi @cdwilson !

> I pulled the latest docker image, and now I'm getting the same results as your [result.zip](https://github.com/INTI-CMNB/KiBot/files/9525616/result.zip). However, if you open the `plot_pdf.pdf` in that zip, you'll see that the soldermask is not being drawn correctly in the PDF. Here's what I see when I open the PDF:

This seems to be a problem with your PDF viewer, here is `evince` on Linux:

![image](https://user-images.githubusercontent.com/22823627/189319360-8af4741a-2d41-4f90-a5fd-a6254bbf6782.png)

BTW: The PDF is used to generate the PNG, if the PNG is OK, then the PDF must be OK.

> Also, I tried this same docker image on a different design, and I'm getting strange results where only part of the soldermask is being drawn:

I don't know which options did you use, I tried kibot 1.3.0 and the PCB you sent me and got:

![ModST_Mouse-assembly 02](https://user-images.githubusercontent.com/22823627/189320533-ee1e44fe-2c1f-464d-9797-87656bc6ae69.png)
![ModST_Mouse-assembly 01](https://user-images.githubusercontent.com/22823627/189320538-51a68bad-0091-4325-a1ce-58b1337e13c6.png)

Are you using the default DPI? I know some of the tools limits the used resources.

**Reply by cdwilson** [reply]
Hmm... you are right, different PDF viewers seem to give different results.

Preview on macOS gives:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189399028-a108fc61-3940-43e0-a54b-cf08b3b43a34.png">

while Adobe Acrobat on macOS gives this for the same file:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189399160-d8e50080-8b00-4293-96bb-2e926b272b03.png">

However, on the other design, the PDF viewer still shows some issues, even on `evince` on Ubuntu:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189399364-a410e690-9363-4d6a-ab18-a006f6a09727.png">


> Are you using the default DPI?

I get the same issue with default or custom DPI (I tried values 72 - 1200).

> I know some of the tools limits the used resources.

Which tools?  Any ideas for things I can try to change or tweak?  Are there any timeouts or things I can try increasing in case some process is not having enough time to complete?

I'll keep experimenting with this to see if I can figure out what is causing this on my side.

**Reply by set-soft** [reply]
> I get the same issue with default or custom DPI (I tried values 72 - 1200).

Did you try using just 1 page?

> Which tools? Any ideas for things I can try to change or tweak? Are there any timeouts or things I can try increasing in case some process is not having enough time to complete?

Doesn't look like a time issue, more like a resources issue.

**Reply by cdwilson** [reply]
Hi @set-soft, I think I am starting to get to the bottom of this.  I think there are two separate issues happening here:

### Issue No. 1
On the `ModST_Mouse.kicad_pcb` board that I sent you earlier, I think the SVG generated by `pcbdraw` is getting clipped incorrectly (not according to the PCB outline).  Since the SVG is now used to generate the other outputs, all the downstream outputs (PDF, PNG, etc) are affected.

Here's a screenshot of the SVG opened in Inkscape:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189452786-998e16d1-d35f-47c3-8d5d-b018a60645e8.png">

However, `pcbdraw` has a command line flag `-e, --adjust-to-edge` which says it will "Adjust size to the board edge".  I tried passing this option when KiBot calls the `pcbdraw` command and it fixed the SVG clipping issue:

```diff
+++ b/kibot/out_pcb_print.py
@@ -776,7 +776,7 @@ class PCB_PrintOptions(VariantOptions):
         pcbdraw_command = self.ensure_tool('PcbDraw')
         # Run PcbDraw to make the heavy work (find the Edge.Cuts path and create masks)
         pcbdraw_file = os.path.join(temp_dir, out_file.replace('.svg', '-pcbdraw.svg'))
-        cmd = [pcbdraw_command, '--no-warn-back', '-f', '']
+        cmd = [pcbdraw_command, '--no-warn-back', '-f', '', '-e']
         if id == B_Mask:
             cmd.append('-b')
         cmd.extend([GS.pcb_file, pcbdraw_file])
```

Here's a screenshot of the same SVG generated with the diff above applied to KiBot:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189452840-610f9181-8fcb-4112-b194-e62b307a7557.png">

This also fixes the downstream PNG, PDF, etc.

### Issue No. 2

The final PDF generated by `pcb_print` is not being rendered correctly by the default PDF viewer (Preview) on macOS, but it is rendered correctly by Adobe Acrobat on macOS.

Here's a screenshot showing the PDF in Preview:

<img width="2524" alt="image" src="https://user-images.githubusercontent.com/14631/189453675-f286bb10-73fa-4c0a-9944-3fa8dfbe47fa.png">

Here's a screenshot showing the same PDF in Adobe Acrobat:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189453734-25cbf919-241f-4278-a0d4-0da0a90078a0.png">

However, when I enable `keep_temporal_files: true`, Preview is able to correctly render the temporary PDF page files generated in the page number (e.g. `01/`) temp directories.  Here's a screenshot showing Preview correctly rendering the realistic soldermask from the temporary PDF file:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/14631/189454025-f81530f9-788e-440e-9613-e45689c612a6.png">

I think the `create_pdf_from_pages` step is doing something to the final PDF that Preview on macOS can't understand/render:
https://github.com/INTI-CMNB/KiBot/blob/877695378c955eec818317aad194a796cc07d538/kibot/out_pcb_print.py#L920

I'm not sure if there is just a bug in Preview, or if the joiner is using some esoteric PDF feature that's not supported, or what... 🤷‍♂️ I'll update if I can figure out what the issue is.

**Reply by cdwilson** [reply]
Hi @set-soft, just a quick update: I'm not exactly sure what changed on my side since yesterday, but I'm no longer seeing the clipping issue (Issue No. 1).  Somehow things just started working today and I have no idea why... (makes me feel like I'm going crazy 🤪)  I'll let you know if I see it happen again, but for now, you can assume that one is resolved.

I'm still having the issue with Preview on macOS, but I don't think that's a KiBot-specific issue.

**Reply by set-soft** [reply]
> I'm still having the issue with Preview on macOS, but I don't think that's a KiBot-specific issue.

Vector graphics are too complex, the SVGs generated by KiCad are technically good, but browsers fails to render them.
I don't know enough about SVGs to try solder mask alternatives.
